File: binding/src/libqtoolbox.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qtoolbox.html
pub const qtoolbox = struct {

    /// New constructs a new QToolBox object.
    pub fn New(parent: ?*anyopaque) QtC.QToolBox {
        return qtc.QToolBox_new(@ptrCast(parent));
    }


    /// New2 constructs a new QToolBox object.
    pub fn New2() QtC.QToolBox {
        return qtc.QToolBox_new2();
    }


    /// New3 constructs a new QToolBox object.
    pub fn New3(parent: ?*anyopaque, f: i64) QtC.QToolBox {
        return qtc.QToolBox_new3(@ptrCast(parent), @intCast(f));
    }


    pub fn Tr(s: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QToolBox_Tr(@ptrCast(s));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qtoolbox.Tr: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn AddItem(self: ?*anyopaque, widget: ?*anyopaque, text: []const u8) i32 {
        const text_str = qtc.libqt_string{
    .len = text.len,
    .data = text.ptr,
};
return qtc.QToolBox_AddItem(@ptrCast(self), @ptrCast(widget), text_str);
    }

    pub fn AddItem2(self: ?*anyopaque, widget: ?*anyopaque, icon: ?*anyopaque, text: []const u8) i32 {
        const text_str = qtc.libqt_string{
    .len = text.len,
    .data = text.ptr,
};
return qtc.QToolBox_AddItem2(@ptrCast(self), @ptrCast(widget), @ptrCast(icon), text_str);
    }

    pub fn InsertItem(self: ?*anyopaque, index: i32, widget: ?*anyopaque, text: []const u8) i32 {
        const text_str = qtc.libqt_string{
    .len = text.len,
    .data = text.ptr,
};
return qtc.QToolBox_InsertItem(@ptrCast(self), index, @ptrCast(widget), text_str);
    }

    pub fn InsertItem2(self: ?*anyopaque, index: i32, widget: ?*anyopaque, icon: ?*anyopaque, text: []const u8) i32 {
        const text_str = qtc.libqt_string{
    .len = text.len,
    .data = text.ptr,
};
return qtc.QToolBox_InsertItem2(@ptrCast(self), index, @ptrCast(widget), @ptrCast(icon), text_str);
    }

    pub fn RemoveItem(self: ?*anyopaque, index: i32) void {
        qtc.QToolBox_RemoveItem(@ptrCast(self), index);
    }

    pub fn SetItemEnabled(self: ?*anyopaque, index: i32, enabled: bool) void {
        qtc.QToolBox_SetItemEnabled(@ptrCast(self), index, enabled);
    }

    pub fn IsItemEnabled(self: ?*anyopaque, index: i32) bool {
        return qtc.QToolBox_IsItemEnabled(@ptrCast(self), index);
    }

    pub fn SetItemText(self: ?*anyopaque, index: i32, text: []const u8) void {
        const text_str = qtc.libqt_string{
    .len = text.len,
    .data = text.ptr,
};
qtc.QToolBox_SetItemText(@ptrCast(self), index, text_str);
    }

    pub fn ItemText(self: ?*anyopaque, index: i32, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QToolBox_ItemText(@ptrCast(self), index);
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qtoolbox.ItemText: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn SetItemIcon(self: ?*anyopaque, index: i32, icon: ?*anyopaque) void {
        qtc.QToolBox_SetItemIcon(@ptrCast(self), index, @ptrCast(icon));
    }

    pub fn ItemIcon(self: ?*anyopaque, index: i32) QtC.QIcon {
        return qtc.QToolBox_ItemIcon(@ptrCast(self), index);
    }

    pub fn SetItemToolTip(self: ?*anyopaque, index: i32, toolTip: []const u8) void {
        const toolTip_str = qtc.libqt_string{
    .len = toolTip.len,
    .data = toolTip.ptr,
};
qtc.QToolBox_SetItemToolTip(@ptrCast(self), index, toolTip_str);
    }

    pub fn ItemToolTip(self: ?*anyopaque, index: i32, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QToolBox_ItemToolTip(@ptrCast(self), index);
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qtoolbox.ItemToolTip: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn CurrentIndex(self: ?*anyopaque, ) i32 {
        return qtc.QToolBox_CurrentIndex(@ptrCast(self));
    }

    pub fn CurrentWidget(self: ?*anyopaque, ) QtC.QWidget {
        return qtc.QToolBox_CurrentWidget(@ptrCast(self));
    }

    pub fn Widget(self: ?*anyopaque, index: i32) QtC.QWidget {
        return qtc.QToolBox_Widget(@ptrCast(self), index);
    }

    pub fn IndexOf(self: ?*anyopaque, widget: ?*anyopaque) i32 {
        return qtc.QToolBox_IndexOf(@ptrCast(self), @ptrCast(widget));
    }

    pub fn Count(self: ?*anyopaque, ) i32 {
        return qtc.QToolBox_Count(@ptrCast(self));
    }

    pub fn SetCurrentIndex(self: ?*anyopaque, index: i32) void {
        qtc.QToolBox_SetCurrentIndex(@ptrCast(self), index);
    }

    pub fn SetCurrentWidget(self: ?*anyopaque, widget: ?*anyopaque) void {
        qtc.QToolBox_SetCurrentWidget(@ptrCast(self), @ptrCast(widget));
    }

    pub fn CurrentChanged(self: ?*anyopaque, index: i32) void {
        qtc.QToolBox_CurrentChanged(@ptrCast(self), index);
    }

    pub fn OnCurrentChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque, i32) callconv(.c) void) void {
        qtc.QToolBox_Connect_CurrentChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn Event(self: ?*anyopaque, e: ?*anyopaque) bool {
        return qtc.QToolBox_Event(@ptrCast(self), @ptrCast(e));
    }

    /// Allows for overriding the related default method
    pub fn OnEvent(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque) callconv(.c) bool) void {
        qtc.QToolBox_OnEvent(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseEvent(self: ?*anyopaque, e: ?*anyopaque) bool {
        return qtc.QToolBox_QBaseEvent(@ptrCast(self), @ptrCast(e));
    }

    pub fn ItemInserted(self: ?*anyopaque, index: i32) void {
        qtc.QToolBox_ItemInserted(@ptrCast(self), index);
    }

    /// Allows for overriding the related default method
    pub fn OnItemInserted(self: ?*anyopaque, callback: *const fn (?*anyopaque, i32) callconv(.c) void) void {
        qtc.QToolBox_OnItemInserted(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseItemInserted(self: ?*anyopaque, index: i32) void {
        qtc.QToolBox_QBaseItemInserted(@ptrCast(self), index);
    }

    pub fn ItemRemoved(self: ?*anyopaque, index: i32) void {
        qtc.QToolBox_ItemRemoved(@ptrCast(self), index);
    }

    /// Allows for overriding the related default method
    pub fn OnItemRemoved(self: ?*anyopaque, callback: *const fn (?*anyopaque, i32) callconv(.c) void) void {
        qtc.QToolBox_OnItemRemoved(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseItemRemoved(self: ?*anyopaque, index: i32) void {
        qtc.QToolBox_QBaseItemRemoved(@ptrCast(self), index);
    }

    pub fn ShowEvent(self: ?*anyopaque, e: ?*anyopaque) void {
        qtc.QToolBox_ShowEvent(@ptrCast(self), @ptrCast(e));
    }

    /// Allows for overriding the related default method
    pub fn OnShowEvent(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque) callconv(.c) void) void {
        qtc.QToolBox_OnShowEvent(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseShowEvent(self: ?*anyopaque, e: ?*anyopaque) void {
        qtc.QToolBox_QBaseShowEvent(@ptrCast(self), @ptrCast(e));
    }

    pub fn ChangeEvent(self: ?*anyopaque, param1: ?*anyopaque) void {
        qtc.QToolBox_ChangeEvent(@ptrCast(self), @ptrCast(param1));
    }

    /// Allows for overriding the related default method
    pub fn OnChangeEvent(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque) callconv(.c) void) void {
        qtc.QToolBox_OnChangeEvent(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseChangeEvent(self: ?*anyopaque, param1: ?*anyopaque) void {
        qtc.QToolBox_QBaseChangeEvent(@ptrCast(self), @ptrCast(param1));
    }

    pub fn Tr2(s: []const u8, c: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QToolBox_Tr2(@ptrCast(s), @ptrCast(c));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qtoolbox.Tr2: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Tr3(s: []const u8, c: []const u8, n: i32, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QToolBox_Tr3(@ptrCast(s), @ptrCast(c), n);
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qtoolbox.Tr3: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QToolBox_Delete(@ptrCast(self));
    }
};

File: binding/src/libqtoolbox.cpp
#include <QEvent>
#include <QFrame>
#include <QIcon>
#include <QShowEvent>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <type_traits>
#include <QToolBox>
#include <QWidget>
#include <qtoolbox.h>
#include "libqtoolbox.h"
#include "libqtoolbox.hxx"

QToolBox* QToolBox_new(QWidget* parent) {
	 return new VirtualQToolBox(parent);
}

QToolBox* QToolBox_new2() {
	 return new VirtualQToolBox();
}

QToolBox* QToolBox_new3(QWidget* parent, int f) {
	 return new VirtualQToolBox(parent, static_cast<QFlags<Qt::WindowType>>(f));
}

libqt_string QToolBox_Tr(const char* s) {
	QString _ret = QToolBox::tr(s);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

int QToolBox_AddItem(QToolBox* self, QWidget* widget, const libqt_string text) {
	return self->addItem(widget, QString::fromUtf8(text.data, text.len));
}

int QToolBox_AddItem2(QToolBox* self, QWidget* widget, const QIcon* icon, const libqt_string text) {
	return self->addItem(widget, *icon, QString::fromUtf8(text.data, text.len));
}

int QToolBox_InsertItem(QToolBox* self, int index, QWidget* widget, const libqt_string text) {
	return self->insertItem(index, widget, QString::fromUtf8(text.data, text.len));
}

int QToolBox_InsertItem2(QToolBox* self, int index, QWidget* widget, const QIcon* icon, const libqt_string text) {
	return self->insertItem(index, widget, *icon, QString::fromUtf8(text.data, text.len));
}

void QToolBox_RemoveItem(QToolBox* self, int index) {
	self->removeItem(index);
}

void QToolBox_SetItemEnabled(QToolBox* self, int index, bool enabled) {
	self->setItemEnabled(index, enabled);
}

bool QToolBox_IsItemEnabled(const QToolBox* self, int index) {
	return self->isItemEnabled(index);
}

void QToolBox_SetItemText(QToolBox* self, int index, const libqt_string text) {
	self->setItemText(index, QString::fromUtf8(text.data, text.len));
}

libqt_string QToolBox_ItemText(const QToolBox* self, int index) {
	QString _ret = self->itemText(index);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QToolBox_SetItemIcon(QToolBox* self, int index, const QIcon* icon) {
	self->setItemIcon(index, *icon);
}

QIcon* QToolBox_ItemIcon(const QToolBox* self, int index) {
	return new QIcon(self->itemIcon(index));
}

void QToolBox_SetItemToolTip(QToolBox* self, int index, const libqt_string toolTip) {
	self->setItemToolTip(index, QString::fromUtf8(toolTip.data, toolTip.len));
}

libqt_string QToolBox_ItemToolTip(const QToolBox* self, int index) {
	QString _ret = self->itemToolTip(index);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

int QToolBox_CurrentIndex(const QToolBox* self) {
	return self->currentIndex();
}

QWidget* QToolBox_CurrentWidget(const QToolBox* self) {
	return self->currentWidget();
}

QWidget* QToolBox_Widget(const QToolBox* self, int index) {
	return self->widget(index);
}

int QToolBox_IndexOf(const QToolBox* self, const QWidget* widget) {
	return self->indexOf(widget);
}

int QToolBox_Count(const QToolBox* self) {
	return self->count();
}

void QToolBox_SetCurrentIndex(QToolBox* self, int index) {
	self->setCurrentIndex(index);
}

void QToolBox_SetCurrentWidget(QToolBox* self, QWidget* widget) {
	self->setCurrentWidget(widget);
}

void QToolBox_CurrentChanged(QToolBox* self, int index) {
	self->currentChanged(index);
}

void QToolBox_Connect_CurrentChanged(QToolBox* self, intptr_t slot) {
    void (*slotFunc)(QToolBox*, int) = reinterpret_cast<void (*)(QToolBox*, int)>(slot);
    QToolBox::connect(self, &QToolBox::currentChanged, [self, slotFunc](int index) {
	slotFunc(self, index);
    });
}

libqt_string QToolBox_Tr2(const char* s, const char* c) {
	QString _ret = QToolBox::tr(s, c);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QToolBox_Tr3(const char* s, const char* c, int n) {
	QString _ret = QToolBox::tr(s, c, n);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QToolBox_Delete(QToolBox* self) {
    delete self;
}

